File: python/mirascope/llm/providers/ollama/provider.py
"""Ollama provider implementation."""

import os
from typing import ClassVar

from openai import AsyncOpenAI, OpenAI

from ..openai.completions.base_provider import BaseOpenAICompletionsProvider


class OllamaProvider(BaseOpenAICompletionsProvider):
    """Provider for Ollama's OpenAI-compatible API.

    Inherits from BaseOpenAICompletionsProvider with Ollama-specific configuration:
    - Uses Ollama's local API endpoint (default: http://localhost:11434/v1/)
    - API key is not required (Ollama ignores API keys)
    - Supports OLLAMA_BASE_URL environment variable

    Usage:
        Register the provider with model ID prefixes you want to use:

        ```python
        import llm

        # Register for ollama models
        llm.register_provider("ollama", "ollama/")

        # Now you can use ollama models directly
        @llm.call("ollama/llama2")
        def my_prompt():
            return [llm.messages.user("Hello!")]
        ```
    """

    id: ClassVar[str] = "ollama"
    default_scope: ClassVar[str | list[str]] = "ollama/"
    default_base_url: ClassVar[str | None] = "http://localhost:11434/v1/"
    api_key_env_var: ClassVar[str] = "OLLAMA_API_KEY"
    api_key_required: ClassVar[bool] = False
    provider_name: ClassVar[str | None] = "Ollama"

    def __init__(
        self,
        *,
        api_key: str | None = None,
        base_url: str | None = None,
    ) -> None:
        """Initialize the Ollama provider.

        Args:
            api_key: API key (optional). Defaults to OLLAMA_API_KEY env var or 'ollama'.
            base_url: Custom base URL. Defaults to OLLAMA_BASE_URL env var
                or http://localhost:11434/v1/.
        """
        resolved_api_key = api_key or os.environ.get(self.api_key_env_var) or "ollama"
        resolved_base_url = (
            base_url or os.environ.get("OLLAMA_BASE_URL") or self.default_base_url
        )

        self.client = OpenAI(
            api_key=resolved_api_key,
            base_url=resolved_base_url,
        )
        self.async_client = AsyncOpenAI(
            api_key=resolved_api_key,
            base_url=resolved_base_url,
        )

    def _model_name(self, model_id: str) -> str:
        """Strip 'ollama/' prefix from model ID for Ollama API."""
        return model_id.removeprefix("ollama/")

File: typescript/src/llm/providers/ollama/provider.ts
/**
 * Ollama provider implementation.
 *
 * Ollama provides an OpenAI-compatible API for local LLM inference.
 */

import { BaseOpenAICompletionsProvider } from "@/llm/providers/openai/completions/base-provider";

const DEFAULT_BASE_URL = "http://localhost:11434/v1/";
const API_KEY_ENV_VAR = "OLLAMA_API_KEY";

export interface OllamaProviderInit {
  apiKey?: string;
  baseURL?: string;
}

/**
 * Provider for Ollama's OpenAI-compatible API.
 *
 * Inherits from BaseOpenAICompletionsProvider with Ollama-specific configuration:
 * - Uses Ollama's local API endpoint (default: http://localhost:11434/v1/)
 * - API key is not required (Ollama ignores API keys)
 * - Supports OLLAMA_BASE_URL environment variable
 *
 * @example
 * ```typescript
 * const provider = new OllamaProvider();
 * const response = await provider.call({
 *   modelId: 'ollama/llama2',
 *   messages: [user('Hello!')],
 * });
 * console.log(response.text());
 * ```
 */
export class OllamaProvider extends BaseOpenAICompletionsProvider {
  readonly id = "ollama" as const;

  /**
   * Create a new Ollama provider instance.
   *
   * @param init - Configuration options
   * @param init.apiKey - API key (defaults to OLLAMA_API_KEY env var, or 'ollama' if not set)
   * @param init.baseURL - Base URL (defaults to OLLAMA_BASE_URL env var, or 'http://localhost:11434/v1/')
   */
  constructor(init: OllamaProviderInit = {}) {
    const resolvedApiKey =
      init.apiKey ?? process.env[API_KEY_ENV_VAR] ?? "ollama";
    const resolvedBaseUrl =
      init.baseURL ?? process.env.OLLAMA_BASE_URL ?? DEFAULT_BASE_URL;

    super({
      id: "ollama",
      apiKey: resolvedApiKey,
      baseURL: resolvedBaseUrl,
    });
  }

  /**
   * Strip 'ollama/' prefix from model ID for Ollama API.
   */
  protected override modelName(modelId: string): string {
    return modelId.startsWith("ollama/") ? modelId.slice(7) : modelId;
  }
}
